### DISCUSSION TITLE
caption only field gone

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#1997] [Ideas & RFCs]
Due to string related decisions we have taken, namely strings will be not quoted, and we automatically go from p-script to f-script parsing mode when we encounter `:` after a non string field, but in string with formatting mode for string chars:

```ftd
-- foo:
x: 1 + 1          ;; x is integer, x gets assigned 2
y: z              ;; y is boolean and gets assigned the value of z
;;                   b is boolean, and this is boolean literal, meaning true / false can not be identifiers. 
;;                   or we keep true as alias for fastn.true, and allow shadowing.
b: true 
s: hello world    ;; s is string
s2: hi, ${ name } ;; format string
```

Coming to caption only field.

In 0.4 we have a feature, that if a record only has one required field, and it is caption, we can pass the value of type of caption to something that is expecting instance of the record, and the record would be automatically constructed. 

```ftd
-- record person:
caption name:
string title: Employee

;; show person wants a person, but we are passing a string
-- show-person: Amit Upadhyay

-- component show-person:
caption person p:

;; .. body omitted ..
-- end: show-person
```

This feature makes code clean in a bunch of cases. You can pass just a string, or you can pass the entire object with the values expected. 

Based on our decisions, we will have to give up on this feature as I am not able to solve the ambiguity here.

**Comment by Arpita-Jaiswal** [comment]
What if a property expects a string, but there’s also a variable with the same name? For example:

```ftd
-- string red: Red is my favourite color.

-- show-color:
favorite: red
```

Question: Should `favorite: red` pass the literal string "red" or the variable `red`?

- *If the parser prioritizes the variable*: How can we explicitly pass "red" as a literal string instead?
- *If the parser prioritizes the literal*: How can we reference the variable `red` instead?

Defining a clear priority and a mechanism to switch between interpretations would resolve this ambiguity.
@amitu

**Reply by amitu** [reply]
What if we do use three modes, `p-script`, `f-script` and `v-script`. The portion after header is `v-script` (value in p-script). 

To refer to a variable you use the `(--) foo: $x` syntax. If you want to do more stuff, e.g., `foo: $x + 1` is allowed. The `$x + 1`  is `v-script`. You can also do `foo: { x + 1 } ` and go to `f-script` mode. In case foo was string, `foo: { .. multi line string ..}`, the content of `{}` would remain string, and it can use `${}`. 

`v-script` starts with `$`, and is not multiline.

**Reply by amitu** [reply]
The multiline string is causing problem, if `foo` is caption only record, the content of `{}` can be either an expression that resolves into that record, or string. Overall it works but it still does not allow caption only record. 

If we say for multiline string you have to use `foo: $" now you can have multi line string"`. If we do this, the `$[]` for multiline string argument gets more weight, as `foo: $[ <multiline ]` looks better (and many strings do contain `"` and they won't have to escaped).
